[AC-4] Add attribution_class related fields to artwork

oxaudo opened the pull request:
https://artsyproduct.atlassian.net/browse/AC-4 requests us to display classification on artwork page. We have multiple strings that need to be associated with each attribution_class value.

Seems like so far the convention was to just hardcode relevant string into files here. In this case all of this looks pretty ugly and repetitive. Definitely open to refactor suggestions here. But so far just gets the values across...

mzikherman commented on src/schema/artwork/index.js:
```diff
@@ -122,6 +122,88 @@ export const artworkFields = () => {
+    attribution_class: {
+      type: GraphQLString,
+      description: "Attribution class as classification name",
+      resolve: artwork => {
```
#minor but the signature can be written as `resolve: ({ attribution_class }) => {` and you can drop the `artwork.` part in the function. Up to you!

mzikherman commented on src/schema/artwork/index.js:
```diff
@@ -122,6 +122,88 @@ export const artworkFields = () => {
+        }
+      },
+    },
+    attribution_class_hover: {
```
I'm thinking the naming here should maybe drop the `_hover` part (since that's a function of the current UI), and maybe be `attribution_class_info` or something?

anandaroop replied:
Agree, maybe `hint` so that it's not tied to a specific kind of device?

oxaudo replied:
👍

mzikherman commented:
One thought looking at the proposed schema, is that we don't have to be limited to mimic-ing the Gravity API response (ie- flat schema/long-ish and potentially repetitive field names).

How about something like:

```
artwork(id: "andy-warhol-skull") {
  attribution {
    class
    info
    description(type: "short")
    description(type: "long")
  }
}
```

You see where I'm going with this, we can be flexible with the schema to group related items together, etc. - take a look and see what you think!

mzikherman commented on src/schema/artwork/index.js:
```diff
@@ -122,6 +122,88 @@ export const artworkFields = () => {
+      description: "Descriptive sentance used as companion for attribution class name display",
+      resolve: artwork => {
+        switch (artwork.attribution_class) {
+          case "editioned multiple": return "High quantity editions, without direct artist involvement."
```
It might be cleaner to have a JSON file here with the info, similar to the spec above.

We already do that for a few things, here's an example file: https://github.com/artsy/metaphysics/blob/master/src/lib/auction_records_trusted.json

Then you can just `require(...)` it.

anandaroop commented on src/schema/artwork/__tests__/artwork.test.js:
```diff
@@ -1,4 +1,5 @@
 import { assign } from "lodash"
+import _ from "lodash"
```
Could be combined with line above — `import { assign, keys }` …

anandaroop commented on src/schema/artwork/index.js:
```diff
@@ -122,6 +122,88 @@ export const artworkFields = () => {
+    },
+    attribution_class_hover: {
+      type: GraphQLString,
+      description: "Descriptive sentance used as companion for attribution class name display",
```
#minor "sentance" &rarr; "sentence", but also maybe it's just a "phrase" not a full sentence

oxaudo commented:
Thanks @mzikherman for recs here.

@anandaroop and myself thought about having this attribution as one object here as opposed to combination of different attribution related fields as you've suggested above. The issue with this (maybe) is that in our case different attributes will be used on different pages. Artwork page will need short and long description and bricks on the rest of the pages will use name and hover. My thinking was that it might be better to only get the attributes needed if they defined as separate attributes.
Not sure it's enough benefit rather than keeping all attribution related stuff in one obj.

anandaroop commented:
Yeah @oxaudo I think we were not thinking "graphql-y" enough when we had that earlier chat. The beauty of it is you only request the fields you need for your client's response, anything from 

```
artwork(id: "cow-55") {
  attribution_class { 
    name
  }
}
```

all the way up to the full enchilada that Matt wrote out above.

oxaudo commented:
@mzikherman, @anandaroop - Attempted to refactor this addressing your comments above. Probably still off track in places, but bear with me:) It's my first pr here...

mzikherman commented on src/schema/types/attribution_class.js:
```diff
@@ -0,0 +1,33 @@
+import {
```
We're a bit inconsistent on where we put these, would you mind moving to `src/schema/artwork` folder? That way attribution and all those other 'partials' are colocated with the `index` which uses all of them.

mzikherman commented on src/schema/types/attribution_class.js:
```diff
@@ -0,0 +1,33 @@
+
+const AttributionClass = new GraphQLObjectType({
+  name: "AttributionClass",
+  description: "Collection of fileds that describe attribution class",
```
Typo: **fields**

mzikherman commented on src/schema/artwork/__tests__/artwork.test.js:
```diff
@@ -937,4 +938,96 @@ describe("Artwork type", () => {
+          expect(data).toEqual({
+            artwork: {
+              attribution_class: {
+                name: attributionClasses[attributionClass].name,
```
In general I think it's preferable to have a more explicit assertion in a spec, as inadvertent `nil` or `undefined` values creeping in can frequently cause tests to still pass (in this style).

Also, I love how each class is tested, but that feels a bit much, maybe we can just have one explicit test for a specific attribution type?

oxaudo commented:
@mzikherman, @anandaroop - Ok, third time is a charm(hopefully)

anandaroop commented on src/lib/attribution_classes.js:
```diff
@@ -0,0 +1,66 @@
+    ].join(" "),
+  },
+  "editioned multiple": {
+    id: "editioned multiple",
```
This id has a space but the one on [L10](https://github.com/artsy/metaphysics/pull/955/files#diff-bd6d53f9c5aff2f8fd7f40cf98028d60R10) has an underscore. Intentional?

Do these need to have parity [with these](https://github.com/artsy/gravity/blob/master/app/models/domain/attribution.rb#L8-L14)?

(Same observation for some other id's in this file)

oxaudo replied:
That was not intentional. Technically we do not even need this Id here, but I though it might be nice to get is as part of the object as something to refer back to in gravity if we need to. Corrected.

anandaroop commented on src/schema/artwork/attribution_class.js:
```diff
@@ -0,0 +1,33 @@
+    },
+    name: {
+      type: GraphQLString,
+      description: "Shortest form of arttribution class display",
```
arttribution &rarr; attribution

anandaroop commented on src/schema/artwork/attribution_class.js:
```diff
@@ -0,0 +1,33 @@
+    },
+    short_description: {
+      type: GraphQLString,
+      description: "Longer version of arttribution class display",
```
Same

anandaroop commented on src/schema/artwork/index.js:
```diff
@@ -29,6 +29,9 @@ import {
 } from "graphql"
+import AttributionClass from "schema/artwork/attribution_class"
+// Mapping of attribution_class ids to AttributionClass values
+import attributionClasses from "../../lib/attribution_classes.js"
```
Not sure what the right pattern for filenames is here — I know there's [active talk about 🐍 vs 🐫 right now](https://artsy.slack.com/archives/C1HH3KNJG/p1519931489000527).

oxaudo replied:
Oh... I did not even realize that there are both versions here in the project:( Just looked at the one above and copied the convention....

oxaudo replied:
Though reading the discussion you linked above those file should be camelCased, no? Did I interpret it correctly?

mzikherman replied:
Yea 👍 on 🐫 !!

mzikherman commented:
Looks great!! 😎 

Aw yea, we got another committer in the house!

![giphy 1](https://user-images.githubusercontent.com/1457859/36914529-a3aa5e8a-1e1b-11e8-89d4-15af06de953e.gif)

anandaroop commented:
Congrats on the Metaphysics PR - nice work!
